### PR TITLE
getProps: centralize 2

### PR DESCRIPTION
**Pull request by hius07**
First approach and discussion in https://github.com/koreader/koreader/pull/10826.

Obtaining document metadata is centralized, deduplicated and optimized.
Also building a base for custom metadata.

(1) For currently opened document we now have in memory `reader.doc_props` extended with `display_title` and customized.

(2) Modules should recevie metadata via centralized `FileManagerBookInfo.getDocProps()`.
If a module already has opened doc or doc_settings, and is getting `doc_props` from them, `FileManagerBookInfo.customizeProps()` should be applied to `doc_props`.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10837)

### CODE REVIEW

**Comment by hius07** [comment]
Changes to readerfooter are just indentation, because of removing check for empty title (`display_title` is always not empty).

File: frontend/document/document.lua [hunk 1]
```diff
@@ -207,8 +207,51 @@ function Document:getNativePageDimensions(pageno)
...
 
-function Document:getProps()
-    return self._document:getDocumentProps()
+function Document:getDocumentProps()
+    -- pdfdocument, djvudocument
+    return self._document:getMetadata()
+    -- credocument, picdocument - overridden by a document implementation
+end
+
+function Document:getProps(props)
```
Comment (poire-z) [inline frontend/document/document.lua:216]: While reading, I find it a bit odd that we could pass a props to this low-level Document method that will actually make it use that and not return the real document props.
I see it is used for documents last opened before 2017 and passing some old `doc_settings:readSetting("stats")` when metadata was saved there.
I guess if we would want to keep that one clean from that, we'd need yet another :getSomethingProps() :) so let's not bother.
Just mentionning that in case you feel the same and would want to think again :)

(Or may be just rename this parameter to something like `use_these_metadata` (a better name than that, but something as explicite)?)

File: frontend/apps/filemanager/filemanagerbookinfo.lua [hunk 2]
```diff
@@ -169,7 +173,38 @@ function BookInfo:show(file, book_props, metadata_updated_caller_callback)
...
+    end
+    return props
+end
+
+-- Returns document metadata (opened document or book (file) metadata or custom metadata).
+function BookInfo.getDocProps(ui, file, book_props, no_open_document, no_customize)
+    local original_props, filepath, display_title
+    if ui then -- currently opened document
+        original_props = ui.doc_settings:readSetting("doc_props")
+        filepath = ui.document.file
```
Comment (poire-z) [inline frontend/apps/filemanager/filemanagerbookinfo.lua:193]: (Dunno if you should use or check `file` when both `ui` and `file` are provided - it feels that when additional arguments are provided, they tweak the behaviour of the previous ones, so if `ui` (generic in its meaning) and `file` (precise) both provided, I would expect file to be used - and `if file == ui.document.file`, use the info from `ui`.)

**Review by poire-z — COMMENTED** [review]
I like all that way better than the previous PR.
A bit hard to follow (and know which to use) with the similarly looking:
`Document:getDocumentProps()` / `_readMetadata()` : raw metadata from engine document
`Document:getProps()` : raw metadata made nil if empty and cleaned up
`BookInfo:getBookProps(file)` metadata from doc_settings or after opening document "file"
`BookInfo:getDocProps(ui, file)` : previous one, overriden with customized stuff, including display_title, and available as ui.doc_props

File: plugins/coverbrowser.koplugin/bookinfomanager.lua [hunk 3]
```diff
@@ -492,37 +492,13 @@ function BookInfoManager:extractBookInfo(filepath, cover_specs)
...
-                    else
-                        -- If the index pattern didn't match (e.g., nothing after the octothorp, or a string),
-                        -- restore the full thing as the series name.
-                        dbrow.series = props.series
-                    end
-                else
-                    dbrow.series = props.series
-                end
+            for k, v in pairs(props) do
+                dbrow[k] = v
```
[the remark this answers is not in the document]
Comment (hius07) [reply plugins/coverbrowser.koplugin/bookinfomanager.lua:500]: I think it is safe, going through `BOOKINFO_COLS_SET` only:
https://github.com/koreader/koreader/blob/21de1a832e5fc5c600aae1b98f6447db3c94dd3b/plugins/coverbrowser.koplugin/bookinfomanager.lua#L568-L570

File: plugins/statistics.koplugin/main.lua [hunk 4]
```diff
@@ -2691,7 +2677,7 @@ function ReaderStatistics:onReadingResumed()
 end
 
 function ReaderStatistics:onReadSettings(config)
-    self.data = config.data.stats or {}
+    self.data = config:readSetting("stats", {})
```
[the remark this answers is not in the document]
Comment (hius07) [reply plugins/statistics.koplugin/main.lua:2680]: I remember an old discussion, at first we did so everywhere, then decided not to do.
In any case, here { } will be filled with the stats, so no empty table saved.

File: frontend/document/document.lua [hunk 5]
```diff
@@ -207,8 +207,51 @@ function Document:getNativePageDimensions(pageno)
...
+    local function makeNilIfEmpty(str)
+        if str == "" then
+            return nil
+        end
```
[the remark this answers is not in the document]
Comment (hius07) [reply frontend/document/document.lua:220]: It is much more effective in the code to have `nil` instead of `""` for empty props.
For the purpose of (future) custom metadata, one or more whitespaces are not an empty prop.

File: frontend/document/document.lua (hunk 1 above)
Comment (hius07) [reply frontend/document/document.lua:216]: Maybe `metadata_to_be_converted`?

File: frontend/apps/filemanager/filemanagerbookinfo.lua (hunk 2 above)
Comment (hius07) [reply frontend/apps/filemanager/filemanagerbookinfo.lua:193]: > when both `ui` and `file` are provided

For now I cannot imagine this case.

File: frontend/document/document.lua (hunk 1 above)
Comment (poire-z) [reply frontend/document/document.lua:216]: It is not really "converted" that much :)
May be `cached_metadata` , `cached_doc_metadata` ?

File: frontend/apps/filemanager/filemanagerbookinfo.lua [hunk 6]
```diff
@@ -187,19 +187,20 @@ end
...
         filepath = file
     end
 
     local props = no_customize and original_props or BookInfo.customizeProps(original_props, filepath)
-    props.display_title = display_title
+    if ui then
+        -- if original title is empty, generate it as filename without extension
+        props.display_title = props.title or filemanagerutil.splitFileNameType(filepath)
+    end
     return props
```
[the remark this answers is not in the document]
Comment (hius07) [reply frontend/apps/filemanager/filemanagerbookinfo.lua:204]: > not called on the currently opened document

That means we call Book information from the file browser on long-pressing on a file.
We agreed that Book information shows original title.

[the remark this answers is not in the document]
Comment (poire-z) [reply frontend/apps/filemanager/filemanagerbookinfo.lua:204]: Still agreeing on that :)
Being not familiar (I already forgot what I reviewed a few days ago :/) with how this is used, I just wondered if there were other cases where this would be called. And you tell me there is none, so ok.
(But even if we don't display `display_title` in "Book information", may be it should be there for consistency and by contract? I wrote after my review:
> BookInfo:getDocProps(ui, file) : previous one, overriden with customized stuff, including display_title, and available as ui.doc_props

)

[the remark this answers is not in the document]
Comment (hius07) [reply frontend/apps/filemanager/filemanagerbookinfo.lua:204]: EDIT: there is one place when we do need `display_title` from the file browser - screensaver showing book title in the message. To be fixed.

File: frontend/apps/filemanager/filemanagerbookinfo.lua [hunk 7]
```diff
@@ -195,13 +197,7 @@ function BookInfo.getDocProps(ui, file, book_props, no_open_document, no_customi
...
         filepath = file
     end
-
-    local props = no_customize and original_props or BookInfo.customizeProps(original_props, filepath)
-    if ui then
-        -- if original title is empty, generate it as filename without extension
-        props.display_title = props.title or filemanagerutil.splitFileNameType(filepath)
-    end
-    return props
+    return no_customize and original_props or BookInfo.customizeProps(original_props, filepath)
```
Comment (poire-z) [inline frontend/apps/filemanager/filemanagerbookinfo.lua:200]: So, it is now BookInfo.customizeProps() that generates display_title - I guess you moved it there to avoid code duplication.
But it feels that it's not really customize'ation :) and should be there even if somewhere we use `no_customize` ?
Anyway, up to you.

Comment (hius07) [reply frontend/apps/filemanager/filemanagerbookinfo.lua:200]: I think further polishing will be done in the next PR with custom metadata.
It's on the way.

**Comment by poire-z** [comment]
> done in the next PR with custom metadata.
It's on the way.

Dunno how you are planning the UX of that.

Just mentionning one situation, if you can give it some thoughts:
dunno how common it is for other people/languages, or if French publisher/users are just careless, but it happens (often enough that this will be my main use of this feature) that in some of my French EPUBs of various origin, the author and title metadata are swapped (that is: the book title is in the author field, the author name appears in the title field, and it is that author name that is displayed in my Statistics Calendar view...).
If I'll be fixing them with this future feature, I'll have to cut & store _elsewhere_ the author field, so I can then cut & paste the title field content in the author field, and then paste that thing I've stored _elsewhere_ in the title field.
So, if you have an idea to solve this (rare?) use case, it would be nice.
(Only thing I have thought of is in your Clipboard dialog, alongside Copy All Copy Line... Paste, there would be a "Swap all with clipboard" (which would 1) copy all into temp variable 2) clear input field 3) paste from clipboard 4) replace clipboard with temp variable), may be hidden and available on long-press on some of the existing button?)

**Comment by hius07** [comment]
A dialog called with a long-press on a property:



<kbd>Set custom</kbd> calls an InputDialog.

**Comment by poire-z** [comment]
The pencils icons there looks nice.
About my use case, I think that if, after customizing one field, I can still view the original (your "View original), **and** I can "Copy it", it would be solved :) (dunno with which widget you will show the original).

**Comment by hius07** [comment]
It will be TextViewer as usual.
So, you'd prefer the Copy button within TextViewer, not the separate button in the dialog?

**Comment by poire-z** [comment]
Oh, sorry, didn't even notice your "Copy original" :/ !
Indeed, this is just perfect :)

**Comment by poire-z** [comment]
> A dialog called with a long-press on a property:

I'm fine with this (I think I prefer this to another widget/menu solution) - but then, the cover display customization feels a bit alone in the top left popup (I thought you would have added something in there :)):
image

**Comment by hius07** [comment]
Got a strong feeling that
https://github.com/koreader/koreader/blob/e7e0d2edb6af95bced95ce88d90a9e08f49fcf7a/frontend/apps/filemanager/filemanagerbookinfo.lua#L371
is an alien in this module, with its `getTextFromPositions` and `getTextBoxes` (despite of the fact that BookInfo is the only beneficiary of these numbers).
The better place seems to be the end of ReaderUI.

**Comment by poire-z** [comment]
Indeed.

> The better place seems to be the end of ReaderUI.

Not sure, ReaderUI is mostly just for bootstrap'ing the reader and its modules.
(OK, there's a `ReaderUI:getCurrentPage()` at its end, but that's just for convenience I guess).

> Or even split it to ReaderRolling and ReaderPaging, there is no duplicated code.

Also not sure: these modules are mostly handling pages, and their respective engines.
(But OK, getCurrentPageLineWordCounts() is about a page :)

I would put it where other getTextFromPositions and getTextBoxes related stuff happens, ReaderHighlight - or even ReaderView as it's about the current view/page.

**Comment by hius07** [comment]
Ready to merge after 08 release.
Custom metadata is ready, based on this PR.

**Review by poire-z — APPROVED** [review]
